File: app/src/main/java/com/mamba/picme/data/download/ParallelFileDownloader.kt
package com.mamba.picme.data.download

import com.mamba.picme.core.common.Logger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.io.IOException
import java.io.RandomAccessFile

/**
 * 大文件分块并行下载器：把单个大文件切成多段，并发 Range 请求下载，按偏移写入同一文件。
 *
 * 用于加速模型中心大文件（如 1.4GB llm.mnn.weight）下载——单连接受 CDN 单流限速，
 * 多段并发可成倍提升吞吐。ModelScope 支持 Range（断点续传已验证）。
 *
 * 各 chunk 以独立 [RandomAccessFile] seek 到非重叠偏移并发写，无需拼接、不双倍占盘。
 */
class ParallelFileDownloader(private val client: OkHttpClient) {

    /**
     * 并发下载 [url] 到 [destFile]（总长 [totalSize]）。
     *
     * @param onBytes 每写入一段字节时回调增量（会被多线程并发调用，调用方需自行保证聚合线程安全）
     * @param isCancelled 调用方取消检测（取消时抛 IOException）
     * @throws IOException 区段 HTTP 非 206/200、响应体空、或最终大小不符
     */
    suspend fun download(
        url: String,
        destFile: File,
        totalSize: Long,
        onBytes: (Long) -> Unit,
        isCancelled: () -> Boolean,
        chunkCount: Int = DEFAULT_CHUNK_COUNT,
        minChunkSize: Long = DEFAULT_MIN_CHUNK_SIZE
    ) {
        val ranges = computeChunkRanges(totalSize, chunkCount, minChunkSize)
        if (ranges.isEmpty()) throw IOException("Invalid totalSize=$totalSize for $url")

        // 预分配文件长度（截断），各 chunk 按偏移并发写入非重叠区段
        withContext(Dispatchers.IO) {
            RandomAccessFile(destFile, "rw").use { it.setLength(totalSize) }
        }
        Logger.i(TAG, "Parallel download: $url -> ${destFile.name}, total=$totalSize, chunks=${ranges.size}")

        coroutineScope {
            for ((index, range) in ranges.withIndex()) {
                launch(Dispatchers.IO) {
                    val req = Request.Builder()
                        .url(url)
                        .header("User-Agent", "PoLang-Android/1.0")
                        .header("Range", "bytes=${range.first}-${range.last}")
                        .build()
                    val call = client.newCall(req)
                    call.execute().use { response ->
                        if (response.code != 206 && response.code != 200) {
                            throw IOException("Chunk $index HTTP ${response.code} for ${destFile.name} (Range 不支持?)")
                        }
                        val body = response.body ?: throw IOException("Empty body, chunk $index ${destFile.name}")
                        body.byteStream().use { input ->
                            RandomAccessFile(destFile, "rw").use { raf ->
                                raf.seek(range.first)
                                val buffer = ByteArray(BUFFER_SIZE)
                                while (true) {
                                    if (isCancelled()) throw IOException("Download cancelled")
                                    val n = input.read(buffer)
                                    if (n == -1) break
                                    raf.write(buffer, 0, n)
                                    onBytes(n.toLong())
                                }
                            }
                        }
                    }
                }
            }
        }

        // 最终大小校验
        val actualLen = destFile.length()
        if (actualLen != totalSize) {
            throw IOException("Size mismatch after parallel download: expected=$totalSize, actual=$actualLen, file=${destFile.name}")
        }
    }

    companion object {
        private const val TAG = "Download"

        /** 单 chunk 读写缓冲 */
        private const val BUFFER_SIZE = 262144 // 256KB

        /** 默认并发区段数（连接数） */
        const val DEFAULT_CHUNK_COUNT = 4

        /** 默认单段最小字节（不足则减少区段数） */
        const val DEFAULT_MIN_CHUNK_SIZE = 8L * 1024 * 1024 // 8MB

        /**
         * 把 [totalSize] 字节切成至多 [chunkCount] 个并发区段，每段不小于 [minChunkSize]。
         *
         * 返回闭区间 LongRange 列表，恰好连续覆盖 [0, totalSize-1]，无重叠无空隙。
         * - totalSize <= 0 → 空
         * - chunkCount <= 1 或 totalSize <= minChunkSize → 单区段 [0, totalSize-1]
         * - 余数均匀分给前若干段，保证各区段长度差 ≤ 1
         */
        fun computeChunkRanges(totalSize: Long, chunkCount: Int, minChunkSize: Long): List<LongRange> {
            if (totalSize <= 0) return emptyList()
            val last = totalSize - 1
            if (chunkCount <= 1 || totalSize <= minChunkSize) return listOf(0L..last)
            val effective = minOf(chunkCount.toLong(), totalSize / minChunkSize).toInt().coerceAtLeast(1)
            if (effective <= 1) return listOf(0L..last)
            val base = totalSize / effective
            val remainder = totalSize % effective
            val ranges = mutableListOf<LongRange>()
            var start = 0L
            for (i in 0 until effective) {
                val len = base + if (i < remainder) 1 else 0
                ranges.add(start..(start + len - 1))
                start += len
            }
            return ranges
        }
    }
}

File: iosApp/PoLang/Platform/ParallelFileDownloader.swift
import Foundation

/// 大文件分块并行下载器（对齐 Android `ParallelFileDownloader`）：
/// 把单个大文件切成多段，并发 Range 请求下载，按偏移写入同一 `.part` 文件。
///
/// 用于加速模型中心大文件（如 1.4GB llm.mnn.weight）——单连接受 CDN 单流限速，
/// 多段并发可提升吞吐。ModelScope 已验证支持 Range。
///
/// 与 Android 差异：支持 **chunk 级断点续传**——完成的 chunk 记录在 `.part.meta`
/// 侧车文件中，暂停/杀进程后重下只补缺失 chunk，且 `.part` 不计入「已下载」
/// （`isModelDownloaded` 只认最终文件名），避免预分配垃圾文件被误判为已下载。
final class ParallelFileDownloader {

    enum DownloadError: Error {
        case invalidTotalSize(Int64)
        case httpError(Int)
        case sizeMismatch(expected: Int64, actual: Int64)
    }

    /// 闭区间字节区段 [start, end]
    struct ChunkRange: Equatable {
        let start: Int64
        let end: Int64
        var length: Int64 { end - start + 1 }
    }

    /// `.part.meta` 侧车：记录已完成的 chunk 下标
    private struct PartMeta: Codable {
        let totalSize: Int64
        let completedChunks: [Int]
    }

    /// 进入分块并行的文件大小阈值（与 Android PARALLEL_DOWNLOAD_THRESHOLD 一致）
    static let parallelThreshold: Int64 = 32 * 1024 * 1024
    /// 默认并发区段数（与 Android DEFAULT_CHUNK_COUNT 一致）
    static let defaultChunkCount = 4
    /// 默认单段最小字节（与 Android DEFAULT_MIN_CHUNK_SIZE 一致）
    static let defaultMinChunkSize: Int64 = 8 * 1024 * 1024

    private let lock = NSLock()
    private var tasks: [URLSessionDownloadTask] = []
    private var completedChunks = Set<Int>()

    /// 取消所有进行中的 chunk 任务（暂停/取消/删除时调用）。
    /// 各 chunk 以 URLError.cancelled 收尾，任务组抛错退出；`.part` + `.meta` 保留供续传。
    func cancelAll() {
        lock.lock()
        let current = tasks
        lock.unlock()
        current.forEach { $0.cancel() }
    }

    /// 并发下载 [url] 到 [partUrl]（总长 [totalSize]），支持 chunk 级续传。
    ///
    /// - Parameter onProgress: 文件维度累计已下载字节（含前次会话完成的 chunk），
    ///   在 delegate 队列触发，调用方自行切线程/节流。
    /// - Throws: 区段 HTTP 非 206/200、字节数不符、或取消（URLError.cancelled）
    func download(url: URL, partUrl: URL, totalSize: Int64,
                  onProgress: @escaping (Int64) -> Void) async throws {
        let ranges = Self.computeChunkRanges(totalSize: totalSize)
        guard !ranges.isEmpty else { throw DownloadError.invalidTotalSize(totalSize) }

        let metaUrl = partUrl.appendingPathExtension("meta")

        // 读取上次进度（totalSize 不一致视为过期，重来）
        if let meta = Self.loadMeta(from: metaUrl), meta.totalSize == totalSize,
           FileManager.default.fileExists(atPath: partUrl.path) {
            completedChunks = Set(meta.completedChunks).intersection(ranges.indices)
        } else {
            try? FileManager.default.removeItem(at: partUrl)
            try? FileManager.default.removeItem(at: metaUrl)
            completedChunks = []
        }

        if !FileManager.default.fileExists(atPath: partUrl.path) {
            FileManager.default.createFile(atPath: partUrl.path, contents: nil)
        }

        // 预分配并截断到 totalSize（对齐 Android RandomAccessFile.setLength）：
        // 各 chunk 按偏移并发写时不再触发空洞扩展，写位置即偏移，且提前暴露磁盘不足。
        // 续传时文件已是 totalSize 长，重复 truncate 无副作用。
        let prealloc = try FileHandle(forWritingTo: partUrl)
        try prealloc.truncate(atOffset: UInt64(totalSize))
        try prealloc.close()

        // 防御性重置（实例正常不复用，但 cancelAll 后 tasks 不残留）
        lock.lock()
        tasks.removeAll()
        lock.unlock()

        let completedBytes = completedChunks.reduce(Int64(0)) { $0 + ranges[$1].length }
        let progress = ChunkProgressTracker(base: completedBytes, onProgress: onProgress)
        if completedBytes > 0 { onProgress(completedBytes) }

        try await withThrowingTaskGroup(of: Void.self) { group in
            for (index, range) in ranges.enumerated() where !completedChunks.contains(index) {
                group.addTask {
                    try await self.downloadChunk(url: url, range: range, index: index,
                                                 partUrl: partUrl, metaUrl: metaUrl,
                                                 totalSize: totalSize, progress: progress)
                }
            }
            try await group.waitForAll()
        }

        // 最终大小校验
        let finalAttrs = try? FileManager.default.attributesOfItem(atPath: partUrl.path)
        let finalSize = (finalAttrs?[.size] as? Int64) ?? 0
        guard finalSize == totalSize else {
            throw DownloadError.sizeMismatch(expected: totalSize, actual: finalSize)
        }
        try? FileManager.default.removeItem(at: metaUrl)
    }

    // MARK: - Chunk

    private func downloadChunk(url: URL, range: ChunkRange, index: Int,
                               partUrl: URL, metaUrl: URL, totalSize: Int64,
                               progress: ChunkProgressTracker) async throws {
        var request = URLRequest(url: url)
        request.setValue("PoLang-iOS/1.0", forHTTPHeaderField: "User-Agent")
        request.setValue("bytes=\(range.start)-\(range.end)", forHTTPHeaderField: "Range")

        let (tempUrl, response) = try await withCheckedThrowingContinuation {
            (cont: CheckedContinuation<(tempFile: URL, response: HTTPURLResponse?), Error>) in
            let task = DownloadTaskHub.shared.startTask(
                with: request,
                onProgress: { written in progress.report(chunk: index, written: written) },
                onCompletion: { cont.resume(with: $0) }
            )
            registerTask(task)
        }
        defer { try? FileManager.default.removeItem(at: tempUrl) }

        let statusCode = response?.statusCode ?? 0
        let tempAttrs = try? FileManager.default.attributesOfItem(atPath: tempUrl.path)
        let tempSize = (tempAttrs?[.size] as? Int64) ?? 0
        // 206 = 正常分段；200 = 服务端忽略 Range 发了整文件（仅单段场景可接受）。
        // 字节数必须恰好等于段长，任何不符都判失败重下。
        guard statusCode == 206 || statusCode == 200, tempSize == range.length else {
            throw DownloadError.httpError(statusCode)
        }

        // 流式写入 part 文件对应偏移（256KB 缓冲，不整段进内存）
        try Self.writeTempFile(at: tempUrl, toPart: partUrl, atOffset: range.start)

        progress.chunkDone(chunk: index, length: range.length)
        markChunkCompleted(index, metaUrl: metaUrl, totalSize: totalSize)
    }

    private func registerTask(_ task: URLSessionDownloadTask) {
        lock.lock()
        tasks.append(task)
        lock.unlock()
    }

    private func markChunkCompleted(_ index: Int, metaUrl: URL, totalSize: Int64) {
        lock.lock()
        completedChunks.insert(index)
        let meta = PartMeta(totalSize: totalSize, completedChunks: completedChunks.sorted())
        if let data = try? JSONEncoder().encode(meta) {
            try? data.write(to: metaUrl, options: .atomic)
        }
        lock.unlock()
    }

    private static func loadMeta(from metaUrl: URL) -> PartMeta? {
        guard let data = try? Data(contentsOf: metaUrl) else { return nil }
        return try? JSONDecoder().decode(PartMeta.self, from: data)
    }

    /// 把临时文件内容流式写入 part 文件的指定偏移
    private static func writeTempFile(at tempUrl: URL, toPart partUrl: URL, atOffset offset: Int64) throws {
        let reader = try FileHandle(forReadingFrom: tempUrl)
        let writer = try FileHandle(forWritingTo: partUrl)
        defer {
            try? reader.close()
            try? writer.close()
        }
        try writer.seek(toOffset: UInt64(offset))
        while true {
            let data = try reader.read(upToCount: 256 * 1024) ?? Data()
            if data.isEmpty { break }
            try writer.write(contentsOf: data)
        }
    }

    // MARK: - Chunk Ranges（对齐 Android computeChunkRanges，纯函数供单测）

    /// 把 [totalSize] 字节切成至多 [chunkCount] 个并发区段，每段不小于 [minChunkSize]。
    /// 返回闭区间列表，恰好连续覆盖 [0, totalSize-1]，无重叠无空隙。
    /// - totalSize <= 0 → 空
    /// - chunkCount <= 1 或 totalSize <= minChunkSize → 单区段
    /// - 余数均匀分给前若干段，保证各区段长度差 ≤ 1
    static func computeChunkRanges(
        totalSize: Int64,
        chunkCount: Int = defaultChunkCount,
        minChunkSize: Int64 = defaultMinChunkSize
    ) -> [ChunkRange] {
        if totalSize <= 0 { return [] }
        let last = totalSize - 1
        if chunkCount <= 1 || totalSize <= minChunkSize {
            return [ChunkRange(start: 0, end: last)]
        }
        let effective = max(1, min(Int64(chunkCount), totalSize / minChunkSize))
        if effective <= 1 {
            return [ChunkRange(start: 0, end: last)]
        }
        let base = totalSize / effective
        let remainder = totalSize % effective
        var ranges: [ChunkRange] = []
        var start: Int64 = 0
        for i in 0..<Int(effective) {
            let len = base + (Int64(i) < remainder ? 1 : 0)
            ranges.append(ChunkRange(start: start, end: start + len - 1))
            start += len
        }
        return ranges
    }
}

// MARK: - 进度聚合（线程安全）

/// 聚合多 chunk 并发进度：base = 已完成 chunk 字节，inFlight = 各 chunk 实时已下载。
private final class ChunkProgressTracker {
    private let lock = NSLock()
    private var base: Int64
    private var inFlight: [Int: Int64] = [:]
    private let onProgress: (Int64) -> Void

    init(base: Int64, onProgress: @escaping (Int64) -> Void) {
        self.base = base
        self.onProgress = onProgress
    }

    func report(chunk: Int, written: Int64) {
        lock.lock()
        inFlight[chunk] = written
        let total = base + inFlight.values.reduce(0, +)
        lock.unlock()
        onProgress(total)
    }

    func chunkDone(chunk: Int, length: Int64) {
        lock.lock()
        inFlight.removeValue(forKey: chunk)
        base += length
        let total = base
        lock.unlock()
        onProgress(total)
    }
}
